### PR TITLE
ABI / codecs: examples, cookbook, pre-release fixes

### PR DESCRIPTION
**Pull request by andreibancioiu**
Added examples in the Cookbook. Minor fixes along the way.

### CODE REVIEW

File: examples/Cookbook.ipynb [hunk 1]
```diff
@@ -716,33 +880,143 @@
...
-    "Create a transaction to upgrade an existing smart contract:"
+    ":::tip\n",
+    "When creating transactions using `SmartContractTransactionsFactory`, even if the ABI is available and provided,\n",
+    "you can still use _typed value_ objects as arguments for deployments and interactions.\n",
+    "\n",
+    "Even further, you can use a mix of _typed value_ objects and plain Python values and objects. For example:\n",
+    "```\n",
+    "args = [U32Value(42), \"hello\", { \"foo\": \"bar\" }, TokenIdentifierValue(\"TEST-abcdef\")];\n",
+    "```\n",
+    ":::"
```
Comment (popenta) [inline examples/Cookbook.ipynb:891]: this is not properly rendered.

Comment (popenta) [reply examples/Cookbook.ipynb:891]: looks good in the docs but not on the notebook.
image

File: examples/Cookbook.ipynb [hunk 2]
```diff
@@ -716,33 +880,143 @@
...
-      "Transaction data: upgradeContract@0061736d0100000001150460037f7f7e017f60027f7f017e60017e0060000002420303656e7611696e74363473746f7261676553746f7265000003656e7610696e74363473746f726167654c6f6164000103656e760b696e74363466696e6973680002030504030303030405017001010105030100020608017f01419088040b072f05066d656d6f7279020004696e6974000309696e6372656d656e7400040964656372656d656e7400050367657400060a8a01041300418088808000410742011080808080001a0b2e01017e4180888080004107418088808000410710818080800042017c22001080808080001a20001082808080000b2e01017e41808880800041074180888080004107108180808000427f7c22001080808080001a20001082808080000b160041808880800041071081808080001082808080000b0b0f01004180080b08434f554e54455200@0506@2a@74657374\n"
+      "SmartContractDeployOutcome(return_code='', return_message='', contracts=[DeployedSmartContract(address='erd1qqqqqqqqqqqqqpgqws44xjx2t056nn79fn29q0rjwfrd3m43396ql35kxy', owner_address='erd1testnlersh4z0wsv8kjx39me4rmnvjkwu8dsaea7ukdvvc9z396qykv7z7', code_hash=b'8Kh\\r\\xf7\\xa9^\\xbc\\xec\\xa0/\\xfb>v\\n/\\xc2\\x88\\xde\\xa1\\xb8\\x02h^\\xf1]\\xf2*\\xe8\\x8b\\xa1[')])\n"
```
Comment (popenta) [inline examples/Cookbook.ipynb:965]: maybe in the future, we override the `__repr__` method to display code hash as hex. just a thought.

File: examples/Cookbook.ipynb [hunk 3]
```diff
@@ -716,33 +880,143 @@
...
      ]
     }
    ],
    "source": [
     "contract_address = Address.from_bech32(\"erd1qqqqqqqqqqqqqpgquzmh78klkqwt0p4rjys0qtp3la07gz4d396qn50nnm\")\n",
```
Comment (popenta) [inline examples/Cookbook.ipynb:1011]: maybe use `new_from_bech32()`

File: examples/Cookbook.ipynb [hunk 4]
```diff
@@ -757,56 +1031,168 @@
...
-      "Transaction data: foo@2a@74657374\n"
+      "Transaction: {'nonce': 0, 'value': '0', 'receiver': 'erd1qqqqqqqqqqqqqpgqws44xjx2t056nn79fn29q0rjwfrd3m43396ql35kxy', 'sender': 'erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th', 'senderUsername': '', 'receiverUsername': '', 'gasPrice': 1000000000, 'gasLimit': 10000000, 'data': 'YWRkQDJh', 'chainID': 'D', 'version': 2, 'options': 0, 'guardian': '', 'signature': '', 'guardianSignature': ''}\n",
+      "Transaction data: add@2a\n"
      ]
     }
    ],
    "source": [
-    "contract_address = Address.from_bech32(\"erd1qqqqqqqqqqqqqpgquzmh78klkqwt0p4rjys0qtp3la07gz4d396qn50nnm\")\n",
+    "contract_address = Address.from_bech32(\"erd1qqqqqqqqqqqqqpgqws44xjx2t056nn79fn29q0rjwfrd3m43396ql35kxy\")\n",
```
Comment (popenta) [inline examples/Cookbook.ipynb:1101]: use `new_from_bech32()`

File: examples/Cookbook.ipynb [hunk 5]
```diff
@@ -757,56 +1031,168 @@
...
+   "source": [
+    ":::tip\n",
+    "When creating transactions using `SmartContractTransactionsFactory`, even if the ABI is available and provided,\n",
+    "you can still use _typed value_ objects as arguments for deployments and interactions.\n",
+    "\n",
+    "Even further, you can use a mix of _typed value_ objects and plain Python values and objects. For example:\n",
+    "```\n",
+    "args = [U32Value(42), \"hello\", { \"foo\": \"bar\" }, TokenIdentifierValue(\"TEST-abcdef\")];\n",
+    "```\n",
+    ":::"
```
Comment (popenta) [inline examples/Cookbook.ipynb:1132]: same comment as above, does not look good in the notebook.

File: examples/Cookbook.ipynb [hunk 6]
```diff
@@ -757,56 +1031,168 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    ":::note\n",
+    "Setting the transaction nonce, signing a transaction and broadcasting it are depicted in a later section.\n",
+    ":::"
```
Comment (popenta) [inline examples/Cookbook.ipynb:1141]: this does not look good either.

File: examples/Cookbook.ipynb [hunk 7]
```diff
@@ -819,17 +1205,39 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Parsing transaction outcome\n",
+    "\n",
+    ":::note\n",
+    "Documentation in this section is preliminary and subject to change.\n",
+    ":::"
```
Comment (popenta) [inline examples/Cookbook.ipynb:1229]: same as above

File: examples/Cookbook.ipynb [hunk 8]
```diff
@@ -819,17 +1205,39 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Decode transaction events\n",
+    "\n",
+    ":::note\n",
+    "Documentation in this section is preliminary and subject to change.\n",
+    ":::"
```
Comment (popenta) [inline examples/Cookbook.ipynb:1240]: same as above

File: examples/Cookbook.ipynb (hunk 4 above)
Comment (andreibancioiu) [reply examples/Cookbook.ipynb:1101]: Fixed.

File: examples/Cookbook.ipynb (hunk 3 above)
Comment (andreibancioiu) [reply examples/Cookbook.ipynb:1011]: Fixed.

File: examples/Cookbook.ipynb (hunk 1 above)
Comment (andreibancioiu) [reply examples/Cookbook.ipynb]: Removed. We'll separately maintain these markers in mx-docs.

**Review by popenta — APPROVED** [review]

File: examples/Cookbook.ipynb (hunk 2 above)
Comment (andreibancioiu) [reply examples/Cookbook.ipynb]: Fixed.

**Review by popenta — APPROVED** [review]

**Review by axenteoctavian — APPROVED** [review]

**Pull request merged**